core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala

```diff
@@ -126,13 +130,28 @@ private[deploy] class HadoopFSDelegationTokenProvider
       Try {
         val newExpiration = token.renew(hadoopConf)
         val identifier = token.decodeIdentifier().asInstanceOf[AbstractDelegationTokenIdentifier]
-        val interval = newExpiration - identifier.getIssueDate
-        logInfo(s"Renewal interval is $interval for token ${token.getKind.toString}")
+        val tokenKind = token.getKind.toString
+        val interval = newExpiration - getIssueDate(tokenKind, identifier)
+        logInfo(s"Renewal interval is $interval for token $tokenKind")
         interval
       }.toOption
     }
     if (renewIntervals.isEmpty) None else Some(renewIntervals.min)
   }
+
+  private def getIssueDate(kind: String, identifier: AbstractDelegationTokenIdentifier): Long = {
```

Comment:
Since others asked for tests maybe we can make this package private and add positive and negative cases (hitting both side of the if condition). I personally don't insist to this.

Reply:
The reason I was negative on adding test is, as we know how thing was broken, we basically expect the test to simulate it and verify it's no longer failing. e.g. testing `obtainDelegationTokens()`. `obtainDelegationTokens()` calls various private methods which call Hadoop API, hence requiring mocks for everything. This is actually an existing debt, and non-trivial to pay off, compared to what I did for the fix.

I'm totally OK to add test verifying this method, as it only requires me to mock AbstractDelegationTokenIdentifier. I'm not sure this fits the needs of adding test, though.

Reply:
That said I don't insist on the test because I agree with you. Just wanted to come up with some coverage. Creating a normal test with security manager + KDC + providers + hadoop sever would require a brain surgery complicated code (I've tried it before in the Kafka area but it was not a success story).

@dongjoon-hyun WDYT?

Reply:
You don't need a security manager for fetching DTs of filesystems or renewing. Example: the S3A ones can be issued if you have AWS creds. It's just that all code written today assumes that filesystems & services only issue DTs if user is kerberized or was launched with an existing set of creds. Example: how to collect a DT to talk to the landsat bucket for the next 12 hours

```
bin/hdfs fetchdt -D fs.s3a.delegation.token.binding=org.apache.hadoop.fs.s3a.auth.delegation.SessionTokenBinding --webservice s3a://landsat-pds/ secrets.bin
```

Renewal is a different bit of work tho'. It would get complicated fast

Reply:
What I've learned from past things is depending on external executables is not always a good decision. This tool must be available on all OS and must behave the exact same way otherwise the test will be pain in the ass.

Reply:
Maybe this is the case here but I'm always a little afraid when this pops up.

Reply:
This discussion shows how hard to deal with test obtainDelegationTokens(). I hope any hero jumps in and deals with this.

Another approach would be changing all private methods to protected methods and fake via subclassing (say, not calling Hadoop API whenever possible), but we lose the ability to test the original methods we override, as well as we modify the code just because we need to add tests. A bit odd rationalization.

Reply:
There is a pretty good reason behind why only sub-parts of token providers are tested, please see Kafka for example...
I still say test getIssueDate.

Reply:
If you still need the pain maybe you can try to mock HadoopFSDelegationTokenProvider.hadoopFSsToAccess with powermock and generate a faulty token. That could lead far.

Reply:
```
  private def getIssueDate(kind: String, identifier: AbstractDelegationTokenIdentifier): Long = {
    val issueDate = identifier.getIssueDate
    if (issueDate > 0L) {
      issueDate
    } else {
      val now = System.currentTimeMillis()
      logWarning(s"Token $kind has not set up issue date properly. (provided: $issueDate) " +
        s"Using current timestamp ($now) as issue date instead. Consult token implementor to fix " +
        "the behavior.")
      now
    }
  }
```

This doesn't seem to be something we would really want to have tests. Unless we separate out the calculation logic and try to craft tests, I don't see any option to have tests for this class.